pdl-compiler/src/backends/rust/mod.rs

```diff
@@ -179,6 +223,206 @@ fn implements_copy(scope: &analyzer::Scope<'_>, field: &ast::Field) -> bool {
+/// packet that can be used to identify with
+/// _certainty_ the child packet.
+///
+/// The discriminant information is:
+///     - field values
+///     - payload size, to disambiguate between children of
+///       identical constant size
+///
+/// The generator will raise warnings if ambiguities remain after all
+/// information is taken into account, i.e. two child packets map to the same
+/// constraints. In this case ambiguities are resolved by trying each child
+/// in order of declaration.
+fn generate_specialize_impl(
+    scope: &analyzer::Scope<'_>,
+    schema: &analyzer::Schema,
+    decl: &ast::Decl,
+    id: &str,
+    data_fields: &[&ast::Field],
+) -> Result<proc_macro2::TokenStream, String> {
+    #[derive(PartialEq, Eq)]
+    struct SpecializeCase {
+        id: String,
+        constraints: HashMap<String, ConstraintValue>,
+        size: analyzer::Size,
+    }
+
+    fn gather_specialize_cases(
+        scope: &analyzer::Scope<'_>,
+        schema: &analyzer::Schema,
+        id: &str,
+        decl: &ast::Decl,
+        data_fields: &[&ast::Field],
+        constraints: &HashMap<String, ConstraintValue>,
+        specialize_cases: &mut Vec<SpecializeCase>,
+    ) {
+        // Add local constraints to the context.
+        let mut constraints = constraints.clone();
+        for c in decl.constraints() {
+            if data_fields.iter().any(|f| f.id() == Some(&c.id)) {
+                constraints.insert(c.id.to_owned(), constraint_value_ast(data_fields, c));
+            }
+        }
+
+        // Generate specialize cases for the child declarations.
+        for decl in scope.iter_children(decl) {
+            gather_specialize_cases(
+                scope,
+                schema,
+                id,
+                decl,
+                data_fields,
+                &constraints,
+                specialize_cases,
+            );
+        }
+
+        // Add a case for the current declaration.
+        specialize_cases.push(SpecializeCase {
+            id: id.to_owned(),
+            constraints,
+            size: schema.decl_size(decl.key) + schema.payload_size(decl.key),
+        });
+    }
+
+    // Create match cases for each child declaration: the union of
+    // tuple of constaint values and packet sizes that will specialize to this
+    // declaration.
+    let mut specialize_cases = Vec::new();
+    for child_decl in scope.iter_children(decl) {
+        gather_specialize_cases(
+            scope,
+            schema,
+            child_decl.id().unwrap(),
+            child_decl,
+            data_fields,
+            &HashMap::new(),
+            &mut specialize_cases,
+        )
+    }
+
+    // List the identifiers of fields constituting the
+    // discriminant tuple.
+    let ids = specialize_cases
+        .iter()
+        .flat_map(|case| case.constraints.keys())
+        .cloned()
+        .collect::<BTreeSet<String>>()
+        .into_iter()
+        .collect::<Vec<String>>();
```

Comment:
This "double collect" is to a) elimintate duplicates and b) sort the strings? It feels like you could skip the second collect, but perhaps you like the simpler `&[String]` signatures better below?

Reply:
Precisely :)